explainer.md

```diff
@@ -186,225 +181,236 @@ https://user-images.githubusercontent.com/93594/141100217-ba1fa157-cd79-4a9d-b3b
+`beforepageshow` is a new event which fires just before the page is shown. Unfortunately `pageshow` fires a long after the page is shown, after `window.onload`. The developer will have to add a listener to this somewhere that's executed before the page is shown.
 
-At the start of the transition, the browser could gather all the `Animation`s active on the stage, and assume the animation is complete once all animations finish.
+- `event.previousURL` - URL of the previous page. Is `null` if cross-origin, or if there was no previous page.
+- `event.direction` - `forward` if moving forward in history, `back` is moving back in history, `new` otherwise.
+- `event.transition` - `null` if Page-A did not offer any elements.
+- `event.transition.data` - Data set by Page-A via `setData`.
+- `event.transition.abandon()` - Abandon the transition.
+- `event.transition.setElement(element, tag, options)` - Same API as in Page-A.
+- `event.transition.ignoreCSSTaggedElements()` - Same API as in Page-A.
 
-In addition, the JS API could include a way to override this by letting the developer provide a promise which keeps the transition active, allowing for animations driven some other way, such as `requestAnimationFrame`.
+In future, this API could include:
 
-This is discussed in [#64](https://github.com/WICG/shared-element-transitions/issues/64).
+- A way to delay the start of the transition, eg to wait for an image to load.
+- A way to enumerate the pseudo elements, which will depend on [CSSPseudoElement](https://drafts.csswg.org/css-pseudo-4/#CSSPseudoElement-interface).
+- A way to manually associate Page-A and Page-B elements, to change the default.
 
 ## Example
 
 Using the sketches above, here's how the example Page-A to Page-B transition could be done:
 
 https://user-images.githubusercontent.com/93594/141100217-ba1fa157-cd79-4a9d-b3b4-67484d3c7dbf.mp4
 
-In Page-A:
-
-```css
-.header {
-  page-transition-tag: header;
-  page-transition-capture: container-and-child;
-}
+TODO
 
-.share-button {
-  page-transition-tag: share-button;
-}
-```
+# SPA API
 
-In Page-B:
+The mechanism for cross-document transitions and SPA transitions involves the same phases, so an SPA API will expose those parts in the same page.
 
-```css
-.header {
-  page-transition-tag: header;
-  page-transition-capture: container-and-child;
-}
+```js
+document.createDocumentTransition(async (transition) => {
+  transition.setElement(document.documentElement, "root");
 
-.share-button {
-  page-transition-tag: share-button;
-}
+  await transition.captureAndHold();
 
-/* Slide the roots from right to left */
-@keyframes slide-left {
-  to { transform: translateX(-100%); }
-}
+  await coolFramework.changeTheDOMToPageB();
 
-::page-transition-root-outgoing {
-  animation-name: slide-left;
-}
+  transition.setElement(document.querySelector(".new-message"), "new-message");
 
-::page-transition-root-incoming {
-  left: 100%;
-  animation-name: slide-left;
-}
+  await transition.start();
```

Comment:
probably don't need to await it

Reply:
It depends when the elements are created. Happy to make it sync if possible.

Reply:
Oh I see, is that what "ready" is for in the MPA version? I do like the idea of a "finalize" signal. The developer indicating that all shared elements on the new page have been specified and elements can now be paired.

But that can just be a sync API, so transition.start() can be sync.